### PR TITLE
avm2: Move DisplayObject assignment into an initializer

### PR DESCRIPTION
**Pull request by Aaron1011**

### CODE REVIEW

File: core/src/avm2/globals/flash/display/shape.rs [hunk 1]
```diff
@@ -1,11 +1,47 @@
...
+
+            return initialize_for_allocator(activation, child, orig_class);
+        }
+        class_object = class.superclass_object();
+    }
+    unreachable!("A Shape subclass should have Shape in superclass chain");
+}
 
 /// Implements `flash.display.Shape`'s 'init' method, which is called from the constructor
 pub fn init<'gc>(
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/shape.rs:47]: This init method can now be entirely removed.

File: core/src/avm2/globals/flash/utils/byte_array.rs [hunk 2]
```diff
@@ -20,27 +19,6 @@ pub fn init<'gc>(
 ) -> Result<Value<'gc>, Error<'gc>> {
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/utils/byte_array.rs:19]: This init method can now be entirely removed.

**Comment by Lord-McSweeney** [comment]
Progresses #11051.

File: core/src/player.rs [hunk 3]
```diff
@@ -2342,9 +2342,9 @@ impl PlayerBuilder {
         let mut player_lock = player.lock().unwrap();
         player_lock.mutate_with_update_context(|context| {
             // Instantiate an empty root before the main movie loads.
-            let fake_root = MovieClip::new(fake_movie, context.gc_context);
-            fake_root.post_instantiation(context, None, Instantiator::Movie, false);
-            context.stage.replace_at_depth(context, fake_root.into(), 0);
+            //let fake_root = MovieClip::new(fake_movie, context.gc_context);
```
Comment (Lord-McSweeney) [inline core/src/player.rs:2345]: Are these changes intentional, or left over from a previous PR?

Comment (Aaron1011) [reply core/src/player.rs:2345]: This is https://github.com/ruffle-rs/ruffle/pull/11083

File: core/src/avm2/globals/flash/display/shape.rs [hunk 4]
```diff
@@ -15,12 +51,6 @@ pub fn init<'gc>(
 ) -> Result<Value<'gc>, Error<'gc>> {
     if let Some(this) = this {
         activation.super_init(this, &[])?;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/shape.rs:53]: This init method can be removed now, right?

File: core/src/avm2/globals/flash/text/text_field.rs [hunk 5]
```diff
@@ -11,6 +12,41 @@ use crate::string::AvmString;
...
+
+            return initialize_for_allocator(activation, child, orig_class);
+        }
+        class_object = class.superclass_object();
+    }
+    unreachable!("A TextField subclass should have TextField in superclass chain");
+}
+
 /// Implements `flash.text.TextField`'s `init` method, which is called from the constructor.
 pub fn init<'gc>(
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/text_field.rs:51]: This init method can be removed now.

File: core/src/avm2/object/bytearray_object.rs [hunk 6]
```diff
@@ -14,14 +15,34 @@ pub fn byte_array_allocator<'gc>(
...
+        .context
+        .library
+        .avm2_class_registry()
+        .class_symbol(class)
+    {
+        if let Some(lib) = activation.context.library.library_for_movie(movie) {
+            if let Some(Character::BinaryData(binary_data)) = lib.character_by_id(id) {
+                Some(ByteArrayStorage::from_vec(binary_data.as_ref().to_vec()))
+            } else {
+                None
```
Comment (Lord-McSweeney) [inline core/src/avm2/object/bytearray_object.rs:28]: This else is still reachable: for example, a class that extends `ByteArray` but doesn't correspond to a `BinaryData` in the symbol class.

Comment (Aaron1011) [reply core/src/avm2/object/bytearray_object.rs:28]: This only occurs if we're successfully looked up the `class_symbol`.

File: core/src/player.rs (hunk 3 above)
Comment (Aaron1011) [reply core/src/player.rs]: My changes previously needed this change, but with the latest version of this PR, all of the tests pass with the fake root in place. I've re-added this code.

File: core/src/avm2/globals/flash/display/display_object.rs [hunk 7]
```diff
@@ -21,7 +20,43 @@ use crate::vminterface::Instantiator;
...
+use swf::Twips;
+use swf::{BlendMode, Rectangle};
+
+pub fn display_object_allocator<'gc>(
+    class: ClassObject<'gc>,
+    _activation: &mut Activation<'_, 'gc>,
+) -> Result<Object<'gc>, Error<'gc>> {
+    // todo we want an Err here
+    // as a script can do `new DisplayObject()`
+    panic!("This allocator should not be called! {:?}", class);
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/display_object.rs:32]: Error here: `ArgumentError: Error #2012: DisplayObject$ class cannot be instantiated.`

File: core/src/avm2/globals/flash/display/simple_button.rs [hunk 8]
```diff
@@ -22,31 +62,39 @@ pub fn init<'gc>(
     if let Some(this) = this {
         activation.super_init(this, &[])?;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/simple_button.rs:63]: This use of `super_init` can be removed.

File: core/src/avm2/globals/flash/display/bitmap.rs [hunk 9]
```diff
@@ -12,13 +13,68 @@ use crate::character::Character;
...
+
 /// Implements `flash.display.Bitmap`'s `init` method, which is called from the constructor
 pub fn init<'gc>(
     activation: &mut Activation<'_, 'gc>,
     this: Option<Object<'gc>>,
     args: &[Value<'gc>],
 ) -> Result<Value<'gc>, Error<'gc>> {
-    if let Some(mut this) = this {
+    if let Some(this) = this {
         activation.super_init(this, &[])?;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/bitmap.rs:78]: This use of `super_init` can now be removed.

File: core/src/avm2/globals/flash/display/display_object.rs (hunk 7 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/display/display_object.rs:32]: This should never be hit - the actual error gets thrown from `DisplayObject.as`

Comment (Lord-McSweeney) [reply core/src/avm2/globals/flash/display/display_object.rs:32]: This can still be hit by a class that extends DisplayObject, which is unintentional. See this SWF for example, which panics with this PR but runs fine (no errors) on master, logging `[object DO]`:
[avm2-tests.swf.zip](https://github.com/ruffle-rs/ruffle/files/11521182/avm2-tests.swf.zip)

Both this PR and master's state is incorrect, though- in FP, it errors with `ArgumentError: Error #2012: DO$ class cannot be instantiated.`

Comment (Aaron1011) [reply core/src/avm2/globals/flash/display/display_object.rs:32]: Good catch - I've added a fix and a test case.

File: core/src/avm2/globals/flash/display/display_object.rs [hunk 10]
```diff
@@ -21,7 +20,51 @@ use crate::vminterface::Instantiator;
...
+
+pub fn display_object_allocator<'gc>(
+    class: ClassObject<'gc>,
+    activation: &mut Activation<'_, 'gc>,
+) -> Result<Object<'gc>, Error<'gc>> {
+    let class_name = class
+        .inner_class_definition()
+        .read()
+        .name()
+        .to_qualified_name(activation.context.gc_context);
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/display_object.rs:34]: This is wrong because it prints the fully qualified class name: for example, in Ruffle, `new DisplayObject()` prints `ArgumentError: Error #2012: flash.display:DisplayObject$ class cannot be instantiated.`, while in Flash it prints `ArgumentError: Error #2012: DisplayObject$ class cannot be instantiated.`
(it's probably worth it adding a test that checeks the error message for `new DisplayObject()`.)

File: core/src/avm2/object/bitmapdata_object.rs [hunk 11]
```diff
@@ -52,7 +52,16 @@ pub struct BitmapDataObjectData<'gc> {
...
+    // Constructs a BitmapData object from a BitmapDataWrapper.
+    // This is *not* used when explicitly constructing a BitmapData
+    // instance from ActionScript (e.g. `new BitmapData(100, 100)`,
+    // or `new MyBitmapDataSubclass(100, 100)`).
+    //
+    // Instead, this is used when constructing a `Bitmap` object,
+    // (from ActionScript or from the timeline), or when we need
+    // to produce a new BitmapData object from a `BitmapData` method
+    // like `clone()`
+    pub fn from_bitmap_data_interal(
```
Comment (Lord-McSweeney) [inline core/src/avm2/object/bitmapdata_object.rs:64]: Typo: "internal" instead of "interal"?

File: core/src/avm2/globals/flash/display/display_object.rs (hunk 10 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/display/display_object.rs]: Fixed

File: core/src/avm2/object/bitmapdata_object.rs (hunk 11 above)
Comment (Aaron1011) [reply core/src/avm2/object/bitmapdata_object.rs]: Fixed

File: core/src/avm2/globals/flash/text/static_text.rs [hunk 12]
```diff
@@ -1,17 +1,18 @@
-use crate::avm2::{Activation, Error, Object, Value};
+use crate::avm2::error::argument_error;
+use crate::avm2::{Activation, ClassObject, Error, Object, Value};
 use crate::avm2_stub_getter;
 
-pub fn native_instance_init<'gc>(
+pub fn static_text_allocator<'gc>(
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/static_text.rs:5]: Won't a class extending `StaticText` hit this error when it tries to construct? IIRC that was the point of having a native instance initializer.

Comment (Aaron1011) [reply core/src/avm2/globals/flash/text/static_text.rs:5]: `StaticText` is final, so that can't happen.

File: core/src/avm2/globals/flash/text/StaticText.as [hunk 13]
```diff
@@ -1,11 +1,8 @@
 package flash.text {
     import flash.display.DisplayObject;
     
-    [Ruffle(NativeInstanceInit)]
+    [Ruffle(InstanceAllocator)]
     public class StaticText extends DisplayObject {
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/StaticText.as:5]: Note: `StaticText` should be a final class, should this be changed here?

**Review by Dinnerbone — APPROVED** [review]

File: core/src/avm2/globals/flash/text/StaticText.as (hunk 13 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/text/StaticText.as]: Done

File: core/src/avm2/globals/flash/geom/sniper_team_notes [hunk 14]
```diff
@@ -0,0 +1,179 @@
+Glow filter: InterfaceDeFreezeAnimationMC
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/geom/sniper_team_notes:1]: Accidentally added?

File: core/src/avm2/globals/flash/media/sound.rs [hunk 15]
```diff
@@ -19,8 +19,6 @@ pub fn init<'gc>(
     args: &[Value<'gc>],
 ) -> Result<Value<'gc>, Error<'gc>> {
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/media/sound.rs:20]: Should Sound also be moved to an instance allocator?

**Comment by SN902** [comment]
I found regression, this commit brings back flickering buttons/sliders in #10208 and [PortageCS5.zip](https://github.com/ruffle-rs/ruffle/files/11525396/PortageCS5.zip)

File: core/src/avm2/globals/flash/media/sound.rs (hunk 15 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/media/sound.rs:20]: I added a test - it looks like Sound is a 'normal' object (the initialization happens during the `super()` call, not during the allocator).

**Review by kmeisthax — COMMENTED** [review]
I'm not sure why `super_init` got removed from `Sound` and `ApplicationDomain` - is there some subtlety of initialization I'm missing? I thought they were "normal" (i.e. no native-allocated fields). I try to keep `super_init` in native code since the AS3 compiler automatically inserts `constructsuper` everywhere anyway, but this isn't particularly important anymore.

The regression @SN902 found is worth looking into but I won't be able to do that for a few more days. Looking at the original PR #10734 it looks like you correctly carried forward everything from that, so it SHOULDN'T be broken, which suggests that we might need to move the check somewhere else, or set a flag here that play/goto checks for.

**Comment by Aaron1011** [comment]
@kmeisthax: The 'super_init' removes were drive by fixes. Those calls were in `init` methods, which are just normal methods called from the constructor (the `super_init` call was left over when we defined the class in actionscript).

I discussed the regression on Discord, but forgot to leave a comment - I fixed it in [`ad15328` (#11084)](https://github.com/ruffle-rs/ruffle/pull/11084/commits/ad153288aaac26eb2d2c1df9e39b5bb7914a55c0)

File: core/src/display_object/graphic.rs [hunk 16]
```diff
@@ -71,10 +71,7 @@ impl<'gc> Graphic<'gc> {
     }
 
     /// Construct an empty `Graphic`.
-    pub fn new_with_avm2(
-        context: &mut UpdateContext<'_, 'gc>,
-        avm2_object: Avm2Object<'gc>,
-    ) -> Self {
+    pub fn new_with_avm2(context: &mut UpdateContext<'_, 'gc>) -> Self {
```
[the remark this answers is not in the document]
Comment (Aaron1011) [reply core/src/display_object/graphic.rs]: I've renamed this to `empty`

File: core/src/display_object/loader_display.rs [hunk 17]
```diff
@@ -36,25 +36,22 @@ impl fmt::Debug for LoaderDisplay<'_> {
...
     movie: Arc<SwfMovie>,
 }
 
 impl<'gc> LoaderDisplay<'gc> {
-    pub fn new_with_avm2(
-        activation: &mut Activation<'_, 'gc>,
-        avm2_object: Avm2Object<'gc>,
-        movie: Arc<SwfMovie>,
-    ) -> Self {
+    pub fn new_with_avm2(activation: &mut Activation<'_, 'gc>, movie: Arc<SwfMovie>) -> Self {
```
[the remark this answers is not in the document]
Comment (Aaron1011) [reply core/src/display_object/loader_display.rs]: I've renamed this to `empty`

**Comment by n0samu** [comment]
With this PR, #8947 is occurring again 😦 
```
ERROR run_frame:run_all_phases_avm2: ruffle_core::avm2::events: Error dispatching event EventObject(EventObject { type: "addedToStage", class: flash.events::Event, ptr: 0x218a7850388 }) to handler FunctionObject(FunctionObject { ptr: 0x218a27e0a88 }) : TypeError: Error #1009: Cannot access a property or method of a null object reference. (accessing field: visible)
        at classes::Level/holderOnAddedToStage()
        at flash.display::DisplayObjectContainer/flash::display::DisplayObjectContainer::addChildAt()
        at classes.app::Main/createLevel()
        at classes.app::Main/nextLevel()
        at classes::ShutterIn/onEnterFrame()
```

**Comment by KrzysztofoGuard** [comment]
Tell me will the game I submitted for revival to the Ruffle team will work or are you already discouraged?  I am asking for a specific answer and I realize that there have been any problems recently.

**Comment by n0samu** [comment]
@KrzysztofoGuard The answer to your question is here https://github.com/ruffle-rs/ruffle/issues/11051#issuecomment-1546905360. There some more things we need to figure out before the game will work at all.

**Comment by waspennator** [comment]
Should close #8409 with pony towers being unable to damage targets.

**Comment by SN902** [comment]
Found regression in [V8MuscleCars.zip](https://github.com/ruffle-rs/ruffle/files/11579989/V8MuscleCars.zip) after intro this game stuck on black screen with ERROR ruffle_core::avm2::events: Error dispatching event EventObject(EventObject { type: "enterFrame", class: flash.events::Event, ptr: 0x2a000207ba8 }) to handler FunctionObject(FunctionObject { ptr: 0x2a07acda048 }) : TypeError: Error #1009: Cannot access a property or method of a null object reference. (accessing field: numChildren)

**Comment by Aaron1011** [comment]
@SN902 I pushed a new commit which should fix the regression

**Comment by SN902** [comment]
> @SN902 I pushed a new commit which should fix the regression
Fixed

**Review by adrian17 — APPROVED** [review]
